src/algorithm/dynamics.hpp

```diff
@@ -106,7 +106,45 @@ namespace pinocchio
     
     return a;
   }
-  
+
+  ///
+  /// \brief Computes the inverse of the constraint matrix [[M JT], [J 0]].
+  /// The matrix is defined when we call forwardDynamics. This method makes use of
+  /// the matrix decompositions performed during the forwardDynamics and returns the inverse.
+  /// The jacobian should be the same that was provided to forwardDynamics.
+  /// Thus you should call forward Dynamics first.
+  /// \param[in] model The model structure of the rigid body system.
+  /// \param[in] data The data structure of the rigid body system.
+  /// \param[out] MJtJ_inv inverse of the MJtJ matrix.
+  ///
+  template<typename Scalar, int Options, template<typename,int> class JointCollectionTpl,
+           typename ConstraintMatrixType>
+  inline void getMJtJInverse(const ModelTpl<Scalar,Options,JointCollectionTpl> &
+                             model,
+                             const DataTpl<Scalar,Options,JointCollectionTpl> & data,
+                             const Eigen::MatrixBase<ConstraintMatrixType> & J,
+                             typename DataTpl<Scalar,Options,JointCollectionTpl>::MatrixXs& MJtJ_inv)
+  {
+    typedef DataTpl<Scalar,Options,JointCollectionTpl> Data;
+    assert(MJtJ_inv.cols() == data.JMinvJt.cols() + model.nv);
+    assert(MJtJ_inv.rows() == data.JMinvJt.rows() + model.nv);
+    typename Data::MatrixXs::Index nc = data.JMinvJt.cols();
+
+    Eigen::Block<typename Data::MatrixXs> topLeft = MJtJ_inv.topLeftCorner(model.nv, model.nv);
+    Eigen::Block<typename Data::MatrixXs> topRight = MJtJ_inv.topRightCorner(model.nv, nc);
+    Eigen::Block<typename Data::MatrixXs> bottomLeft = MJtJ_inv.bottomLeftCorner(nc, model.nv);
+    Eigen::Block<typename Data::MatrixXs> bottomRight = MJtJ_inv.bottomRightCorner(nc, nc); 
+
+    bottomRight.setIdentity(); topLeft.setIdentity();
+    data.llt_JMinvJt.solveInPlace(bottomRight);    cholesky::solve(model, data, topLeft);
+
+    bottomLeft.noalias() = J*topLeft;
+    topRight.noalias() = bottomLeft.transpose() * bottomRight;
+    topLeft.noalias() -= topRight*bottomLeft;
+    bottomLeft = topRight.transpose();
+    bottomRight *=-1;
```

Comment:
You can avoid this last operation to save computation times.

Reply:
I have the choice between
1) saving A; do computations using A; replace A with -A
2) saving -A; do computations using A by calling (-)(-A) everytime.

I chose the first one. Do you think that the second would be better?

Reply:
Yes, Eigen will do the job for you in this precise context!

Reply:
I would still need to do multiply by -1, since I am creating the A matrix by solveInPlace(-Identity)
I could do a *-1 on the diagonal only after setting identity. this would be faster.

Reply:
You can set it to -Identity. Eigen will the optimization for you.